data/fileio/CSVStreamWriter.h

```diff
@@ -0,0 +1,100 @@
+/* -*- c-basic-offset: 4 indent-tabs-mode: nil -*-  vi:set ts=8 sts=4 sw=4: */
+
+/*
+    Sonic Visualiser
+    An audio file viewer and annotation editor.
+    Centre for Digital Music, Queen Mary, University of London.
+    This file copyright 2017 Lucas Thompson.
+    
+    This program is free software; you can redistribute it and/or
+    modify it under the terms of the GNU General Public License as
+    published by the Free Software Foundation; either version 2 of the
+    License, or (at your option) any later version.  See the file
+    COPYING included with this distribution for more information.
+*/
+
+#ifndef _CSV_STREAM_WRITER_H_
+#define _CSV_STREAM_WRITER_H_
+
+#include "base/BaseTypes.h"
+#include "base/Selection.h"
+#include "base/ProgressReporter.h"
+#include "base/DataExportOptions.h"
+#include "data/model/Model.h"
+#include <QString>
+#include <algorithm>
+
+namespace CSV
+{
+using Completed = bool;
+
+template <class OutStream>
+auto writeToStreamInChunks(
+    OutStream& oss,
+    const Model& model,
+    const Selection& extents,
+    ProgressReporter* reporter = nullptr,
+    QString delimiter = ",",
+    DataExportOptions options = DataExportDefaults,
+    const sv_frame_t blockSize = 16384
+) -> Completed
+{
+    if (blockSize <= 0) return false;
+    sv_frame_t readPtr = extents.isEmpty() ?
+        model.getStartFrame() : extents.getStartFrame();
+    sv_frame_t endFrame = extents.isEmpty() ?
+        model.getEndFrame() : extents.getEndFrame();
+    int previousPercentagePoint = 0;
+
+    const auto wasCancelled = [&reporter]() { 
+        return reporter && reporter->wasCancelled(); 
+    };
+
+    while (readPtr < endFrame) {
+        if (wasCancelled()) return false;
+
+        const auto start = readPtr;
+        const auto end = std::min(start + blockSize, endFrame);
+
+        oss << model.toDelimitedDataStringSubsetWithOptions(
+            delimiter,
+            options,
+            start,
+            end
+        ) << (end < endFrame ? "\n" : "");
+
+        const auto currentPercentage = 100 * end / endFrame;
+        const bool hasIncreased = currentPercentage > previousPercentagePoint;
```

Comment:
The percentage calculation is only correct if we started at zero I think? Otherwise we need (end-start) / (endFrame-startFrame) and of course only if the denominator is non-zero. This also suggests that the name `end` is not sufficiently different from `endFrame` -- maybe it could be `blockEnd` or something.

Reply:
Good catch, cheers. A symptom of this is noticeable when using multi selection (which, sadly I had overlooked). When choosing a short region and a large one, it can be observed that the progress bar is wrong for the majority of the export. The resulting CSV file is still actually valid. I'll add a test and fix this.